47 garden ai pipeline create

OwenPriceSkelly opened the pull request:
closes #47 

<!-- readthedocs-preview garden-ai start -->
----
:books: Documentation preview :books:: https://garden-ai--66.org.readthedocs.build/en/66/

<!-- readthedocs-preview garden-ai end -->

OwenPriceSkelly committed:
client methods to read/write from local database

tests for local db helper methods

misc fixes for tests

[fix] beartype.typing import only post-3.8

OwenPriceSkelly committed:
rename app `garden` -> `garden-ai`

(`garden-ai garden create` > `garden garden create`, etc)

OwenPriceSkelly committed:
`garden-ai {noun} {verb}` instead of verb-first

OwenPriceSkelly committed:
creates pipeline with "shortname" identifier,

i.e. a valid python variable name for the scaffolded file.

if one isn't provided, the title is (slugified? snakeified?) reformatted
to be a valid identifier and is used instead.

create pipeline test

OwenPriceSkelly committed:
`poetry add Jinja2`

convince poetry to include templates

OwenPriceSkelly committed:
jinja2 templating code

started templating pipeline

OwenPriceSkelly committed:
formatting (big diff; no functional changes)

test repair 🔧

skip cli test for 3.7 only

the pytest/mocked `call_args.kwargs` needed for the test
backwards-incompatibly changed in 3.8

I think we're going to drop 3.7 anyways, so I don't think we need a
3.7-only version of the CLI tests.

WillEngler commented on templates/pipeline:
```diff
@@ -0,0 +1,50 @@
+## THIS FILE WAS AUTOMATICALLY GENERATED {{  }} ##
```
Why do we have this file as well as garden_ai/templates/pipeline?

OwenPriceSkelly replied:
we shouldn't, just slipped by me -- this is where I'd assumed jinja/poetry would look for templates but they needed to be down a level

WillEngler commented on tests/test_client.py:
```diff
@@ -110,3 +110,38 @@ def test_client_invalid_auth_token(mocker, mock_authorizer_tuple, token, mock_ke
+    mocker.patch("garden_ai.client.LOCAL_STORAGE", new=tmp_path)
+    uuid = garden_all_fields.uuid
+    garden_client.put_local(garden_all_fields)
+    record = garden_client.get_local(uuid)
```
It seems like the existing application code is still writing to {garden_id}.json right now, right? Is it a big lift to convert the code that's persisting models now so these parts aren't out of sync?

WillEngler commented on tests/test_client.py:
```diff
@@ -110,3 +110,38 @@ def test_client_invalid_auth_token(mocker, mock_authorizer_tuple, token, mock_ke
+def test_local_storage_keyerror(mocker, garden_client, garden_all_fields, tmp_path):
+    # mock to replace "~/.garden/db"
+    tmp_path.mkdir(parents=True, exist_ok=True)
+    mocker.patch("garden_ai.client.LOCAL_STORAGE", new=tmp_path)
```
Thinking out loud ... it would be great if our tests weren't doing a lot of filesystem mocking.  Probably not worth rethinking it on this PR, but I think we can get to a place where our tests are just manipulating a dict in memory and we can mock out the `read_from_disk` and `write_to_disk` calls.

WillEngler reviewed (commented):
Looking good! I have some feedback mostly on persistence

OwenPriceSkelly commented on garden_ai/app/pipeline.py:
```diff
@@ -0,0 +1,207 @@
+pipeline_app = typer.Typer(name="pipeline", no_args_is_help=True)
+
+
+def validate_identifier(name: str) -> str:
```
(for posterity after sync review) agreed, the word 'validate' was sticky in my brain because of the way pydantic "validators" work similarly (take a maybe-valid value and return a valid one) but that's very opinionated; even if we share pydantic's "parse, don't validate" mindset we don't need to share the "parse, don't validate; also you do the parse with something called `@validator`" mindset https://github.com/pydantic/pydantic/issues/578

OwenPriceSkelly committed:
delete redundant template

OwenPriceSkelly committed:
rename validate_identifier -> clean_identifier

OwenPriceSkelly committed:
misc refactor following code review

e.g. split local DB logic (`put_local` -> `put_local_{garden,
pipeline}`); corresponding fixes/renames in tests; remove
the "run as a script" bit from template; etc.

WillEngler commented on garden_ai/app/pipeline.py:
```diff
@@ -197,11 +197,11 @@ def dummy_step(arg: object) -> object:
-    client.put_local(pipeline)
     if verbose:
-        metadata = client.get_local(pipeline.uuid)
+        client.put_local_pipeline(pipeline)
```
I think we shouldn't store the pipeline before it's registered.

WillEngler commented on garden_ai/client.py:
```diff
@@ -353,7 +359,41 @@ def register_metadata(self, garden: Garden, out_dir=None):
+        data["gardens"] = gardens
+        contents = json.dumps(data)
+
+        with open(LOCAL_STORAGE / "data.json", "w+") as f:
```
I think we're missing some opportunities to DRY our persistence operations. For example it looks like this block

```
 if (LOCAL_STORAGE / "data.json").exists():
            with open(LOCAL_STORAGE / "data.json", "r+") as f:
                raw_data = f.read()
                if raw_data:
                    data = json.loads(raw_data)
```

is present in put_local_garden and put_local_pipeline.

WillEngler commented on garden_ai/client.py:
```diff
@@ -405,16 +448,50 @@ def put_local(self, obj: Union[Garden, Pipeline]) -> None:
+        """Helper: fetch a pipeline record from 'local database', if one exists.
+
+        Find entry with key matching ``uuid`` and return the associated metadata
+        extracted from ``~/.garden/db/data.json``
```
I think the /db is a typo now

WillEngler reviewed (commented):
I think we should have one place where we read the data file and one place where we write to it.

OwenPriceSkelly committed:
DRY-er local storage methods
